crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1672,6 +1690,80 @@ impl KnownInstanceType {
             _ => None,
         }
     }
+
+    fn member(self, db: &'db dyn Db, name: &str) -> Symbol<'db> {
+        match (self, name) {
+            (Self::TypeVar(typevar), "__name__") => Symbol::Type(
+                Type::StringLiteral(StringLiteralType::new(db, typevar.name(db).as_str())),
+                Boundness::Bound,
+            ),
+            (Self::TypeVar(typevar), "__bound__") => Symbol::Type(
+                typevar
+                    .upper_bound(db)
+                    .map(|ty| ty.to_meta_type(db))
+                    .unwrap_or(KnownClass::NoneType.to_instance(db)),
+                Boundness::Bound,
+            ),
+            (Self::TypeVar(typevar), "__constraints__") => Symbol::Type(
+                Type::Tuple(TupleType::new(
+                    db,
+                    typevar
+                        .constraints(db)
+                        .iter()
+                        .map(|ty| ty.to_meta_type(db))
+                        .collect::<Vec<_>>()
+                        .into_boxed_slice(),
+                )),
+                Boundness::Bound,
+            ),
+            (Self::TypeVar(typevar), "__default__") => Symbol::Type(
+                typevar
+                    .default_ty(db)
+                    .map(|ty| ty.to_meta_type(db))
+                    .unwrap_or(Type::KnownInstance(KnownInstanceType::NoDefault)),
+                Boundness::Bound,
+            ),
+            _ => self.instance_fallback(db).member(db, name),
+        }
+    }
+}
+
+#[salsa::interned]
+pub struct TypeVarInstance<'db> {
+    /// The name of this TypeVar (e.g. `T`)
+    #[return_ref]
+    pub name: ast::name::Name,
+
+    /// The upper bound or constraint on the type of this TypeVar
+    #[return_ref]
+    bound_or_constraints: Option<TypeVarBoundOrConstraints<'db>>,
+
+    /// The default type for this TypeVar
+    pub(crate) default_ty: Option<Type<'db>>,
+}
+
+impl<'db> TypeVarInstance<'db> {
+    pub(crate) fn upper_bound(self, db: &'db dyn Db) -> Option<Type<'db>> {
+        if let Some(TypeVarBoundOrConstraints::UpperBound(ty)) = self.bound_or_constraints(db) {
+            Some(*ty)
+        } else {
+            None
+        }
+    }
+
+    pub(crate) fn constraints(self, db: &'db dyn Db) -> &[Type<'db>] {
+        if let Some(TypeVarBoundOrConstraints::Constraints(tys)) = self.bound_or_constraints(db) {
+            tys
+        } else {
+            &[]
```

Comment:
Given that a list of constraints is interpreted as: *can only ever be solved as being exactly one of the constraints given*, it seems a bit strange to me to map the "no constraints" case to an empty list (which could also mean: can't be solved at all, because there are no options given). I realize that both `mypy` and `pyright` disallow this (*"TypeVar must have at least two constrained types"*), but it still seems potentially dangerous to conflate these two potential meanings. So maybe we should change the return type to `Option<&[Type<'db>]>` and return `None` here?

Reply:
Yes, makes sense; done.